types-core/index.d.ts

```diff
@@ -129,4 +135,386 @@ declare module '@azure/functions-core' {
          */
         dispose(): any;
     }
+
+    /**
+     * Registers the main programming model to be used for a Node.js function app
+     * Only one programming model can be set. The last programming model registered will be used
+     * If not explicitly set, a default programming model included with the worker will be used
+     */
+    function setProgrammingModel<TContext>(programmingModel: ProgrammingModel<TContext>): void;
+
+    /**
+     * Returns the currently registered programming model
+     * If not explicitly set, a default programming model included with the worker will be used
+     */
+    function getProgrammingModel<TContext>(): ProgrammingModel<TContext>;
+
+    /**
+     * A set of information and methods that describe the model for handling a Node.js function app
+     * Currently, this is mainly focused on invocation
+     */
+    interface ProgrammingModel<TContext> {
+        /**
+         * A name for this programming model, generally only used for tracking purposes
+         */
+        name: string;
+
+        /**
+         * A version for this programming model, generally only used for tracking purposes
+         */
+        version: string;
+
+        /**
+         * Returns a new instance of the invocation model for each invocation
+         */
+        getInvocationModel(coreContext: CoreInvocationContext): InvocationModel<TContext>;
+    }
+
+    /**
+     * Basic information and helper methods about an invocation provided from the core worker to the programming model
+     */
+    interface CoreInvocationContext {
+        /**
+         * A guid unique to this invocation
+         */
+        invocationId: string;
+
+        /**
+         * The invocation request received by the worker from the host
+         */
+        request: RpcInvocationRequest;
+
+        /**
+         * Metadata about the function
+         */
+        metadata: RpcFunctionMetadata;
+
+        /**
+         * Describes the current state of invocation, or undefined if between states
+         */
+        state?: InvocationState;
+
+        /**
+         * The recommended way to log information
+         */
+        log(level: RpcLog.Level, category: RpcLog.RpcLogCategory, message: string): void;
```

Comment:
They both do logging stuff. However, the core api expects an already formatted string here (hence why message is just type `string`) and the programming model package is in charge of formatting the string. That's because of bugs like this where people want different formatting behavior that conflict with each other, so it would be a breaking change if we ever modified this: https://github.com/Azure/azure-functions-nodejs-worker/issues/586